### PR TITLE
Add a provider definition for Webex

### PR DESCRIPTION
**Pull request by cyphermox**
Adding a provider for Webex/Webex Teams incoming webhooks support.

Signed-off-by: Mathieu Trudel-Lapierre <mtrudel@stack8.com>

### CODE REVIEW

**Comment by cyphermox** [comment]
I noticed there was an additional tree for "specs". I'm not familiar with ruby per-se. Was that for tests?

File: lib/discourse_chat/provider/webex/webex_provider.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      category = "[#{I18n.t('uncategorized_category_name')}]"
+    elsif topic.category
+      category = (topic.category.parent_category) ? "[#{topic.category.parent_category.name}/#{topic.category.name}]" : "[#{topic.category.name}]"
+    end
+
+    message = {
+      "markdown": "**#{topic.title}**: #{category} #{topic.tags.present? ? topic.tags.map(&:name).join(', ') : ''}(#{post.full_url}) from #{full_name}(#{display_name}): #{post.excerpt(SiteSetting.chat_integration_webex_excerpt_length, text_entities: true, strip_links: true, remap_emoji: true)}"
+    }
+
+    message
```
Comment (davidtaylorhq) [inline lib/discourse_chat/provider/webex/webex_provider.rb:50]: This line is a little long. How about something like
```suggestion
    markdown = "**#{topic.title}**: #{category} "
    markdown += "#{topic.tags.map(&:name).join(', ')} " if topic.tags.present?
    markdown += "(#{post.full_url}) from #{full_name} (#{display_name}) "
    markdown += post.excerpt(SiteSetting.chat_integration_webex_excerpt_length, 
                             text_entities: true, 
                             strip_links: true, 
                             remap_emoji: true
                            )
    { markdown: markdown }
```

**Review by davidtaylorhq — COMMENTED** [review]
This is looking good - thanks @cyphermox. Just one comment on the ruby code.

> I noticed there was an additional tree for "specs". I'm not familiar with ruby per-se. Was that for tests?

Yes exactly - we do need some tests in order to merge this, just to make sure we don't have any regressions in future. It should be possible to copy and adapt the specs from another provider. If you have a discourse development environment set up, you can run the tests like `bin/rake plugin:spec[discourse-chat-integration]`

Ideally we should also add some WebEx specific documentation to [our howto guides on Meta](https://meta.discourse.org/t/chatroom-integration-plugin-discourse-chat-integration/66522). Is that something you'd be able to put together?

File: lib/discourse_chat/provider/webex/webex_provider.rb (hunk 1 above)
Comment (cyphermox) [reply lib/discourse_chat/provider/webex/webex_provider.rb:50]: Sure; I'm up for improving things. I just did the minimum viable so far, to limit how much work time I'd spend on it :)

I'll add the tests and fix the long strings shortly.

Comment (cyphermox) [reply lib/discourse_chat/provider/webex/webex_provider.rb:50]: Do you have guidelines for line length? Should I also adjust the others, and if so; how do you feel about operators -- end of line or beginning of line, etc. For example:

```
    elsif topic.category
      category = (topic.category.parent_category)
               ? "[#{topic.category.parent_category.name}/#{topic.category.name}]"
               : "[#{topic.category.name}]"
```
or
```
      raise ::DiscourseChat::ProviderError.new info: { error_key: error_key,
                                                       request: req.body,
                                                       response_code: response.code,
                                                       response_body: response.body }
```

Is that good?

Comment (davidtaylorhq) [reply lib/discourse_chat/provider/webex/webex_provider.rb:50]: That all looks good to me. We don't have any super strict rules on formatting or line length, so as long as it's readable, it's fine. The rules we do have are enforced by the rubocop config (which looks like it's passing 😄 )

**Comment by cyphermox** [comment]
This is still going to require translation strings. I can do the French ones right away.

I haven't done testing on this code yet.

**Comment by davidtaylorhq** [comment]
All we need for now is English translation strings. Other languages are handled through our Crowdin integration. After this PR is merged, the strings will become available for translation over there. More details at https://meta.discourse.org/t/contribute-a-translation-to-discourse/14882

> I haven't done testing on this code yet.

If you're able to spin up a dev instance and test, that would be great. Some screenshots of how to configure WebEx, and what the notifications look like in WebEx would also be useful for documentation on Meta.

**Comment by cyphermox** [comment]
> All we need for now is English translation strings. Other languages are handled through our Crowdin integration. After this PR is merged, the strings will become available for translation over there. More details at https://meta.discourse.org/t/contribute-a-translation-to-discourse/14882
> 
> > I haven't done testing on this code yet.
> 
> If you're able to spin up a dev instance and test, that would be great. Some screenshots of how to configure WebEx, and what the notifications look like in WebEx would also be useful for documentation on Meta.

Yeah, I just meant that I was going to do that, and hadn't yet. As for the howto, it's already written, I'll just need to publish it. I already have a screenshot of the resulting output; I basically copy-pasted the Microsoft Teams doc.

**Comment by cyphermox** [comment]
Trying to run this from within the docker container I already spun up for my test/production instance, rake fails to properly start rspec. I'm way out of my depth here, it's my first time ever touching any ruby code.

```
Traceback (most recent call last):
	2: from /var/www/discourse/bin/rspec:18:in `<main>'
	1: from /usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/rubygems_integration.rb:413:in `block in replace_bin_path'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/rubygems_integration.rb:374:in `block in replace_bin_path': can't find executable rspec for gem rspec-core. rspec-core is not currently included in the bundle, perhaps you meant to add it to your Gemfile? (Gem::Exception)
rake aborted!
```

Trying to follow recommendations on how to fix this, but I'm not getting much traction:
`bundle update rspec` works fine.
`bundle update rspec-core` works fine.

```
# bundle update rspec-rails
Unable to find a spec satisfying actionpack (= 6.0.3.3) in the set. Perhaps the lockfile is corrupted?
```
`actionpack` at the proper version is written in _Gemfile.lock_; so any suggestion on how to quickly test this the right way?

**Comment by davidtaylorhq** [comment]
If you have a Linux or MacOS environment, the easiest way to get something working is probably this: https://meta.discourse.org/t/beginners-guide-to-install-discourse-for-development-using-docker/102009

RSpec tests are also run as part of the "Plugin Tests / backend" CI here on GitHub - it looks like they're passing.

**Comment by cyphermox** [comment]
> If you have a Linux or MacOS environment, the easiest way to get something working is probably this: https://meta.discourse.org/t/beginners-guide-to-install-discourse-for-development-using-docker/102009
> 

That alone won't work. I'm running into major issues with the test scripts due to UID mismatches.

Starting the docker container as it is done in those scripts only works either because you run things unprivileged (rootless docker) or because in many cases it happens to map the discourse user to the same UID as your own (you get 1000 usually, and discourse is started as a normal user and also gets 1000 inside the container). I'm synced to an external authentication source, so I'll never get UID 1000. I tried a couple of things to circumvent that, but with not much luck. I still have some things I could try, but if we're good with the tests as they were run by the CI scripts, that would probably be best for me.

**Review by davidtaylorhq — APPROVED** [review]
Looks great to me - thanks so much for the contribution @cyphermox ❤️

**Pull request merged**